benchmarks/ref/include/Tensor.h

```diff
@@ -0,0 +1,184 @@
+//===- Tensor.h - ---------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "llvm/ADT/STLExtras.h"
+
+#include <cassert>
+#include <cstring>
+#include <functional>
+#include <initializer_list>
+#include <iostream>
+#include <numeric>
+#include <vector>
+
+namespace {
+using ListArg = std::vector<unsigned>;
+using InitArg = std::initializer_list<unsigned>;
+} // namespace
+
+// Tensor: represents a tensor in memory, basically a pointer to data plus
+// rank and dimension.
+template <typename T> class Tensor {
```

Comment:
Do we need to differentiate between CPU tensor and GPU tensor?

For example, if we want to "read" a result tensor on the GPU, we'll need to copy it back or something. 

Or is that supposed to be done by the kernels themselves? And if so, how can we avoid timing that?

Reply:
Currently, we need a bit due to memory management. That's why there's `CudaTensor` which provides GPU wrapper over other Tensor types.

Otherwise, good questions but I'll worry about them if I ever need these features. For now I don't expect to expand much on the side of manual CUDA kernels.

Reply:
About data movement outside vs inside kernel.
It's a bit of a design choice but we'll have to see what others do when we want to compare any numbers. For now, both in the reference CUDA implementation and MLIR kernels the args are copied to GPU outside of the timing loop.

In the future, it might be beneficial to get more exact numbers using some Nvidia profiler.